### ISSUE TITLE
Bundler fails with Rubygems 2.7.5

### ISSUE THREAD

**Issue by olbrich** [rubygems/rubygems#2180]
I'm having a problem or would like to suggest a feature.

My current problem is 'bundle install fails with rubygems 2.7.5 (ruby 2.3.6) on ubuntu 16.04'

This issue is related to:

  - [ ] Network problems
  - [ ] Installing a library
  - [ ] Publishing a library
  - [ ] The command line `gem`
  - [x] Other

Here are my current environment details:

```
$ gem env version
2.7.5
```

I will abide by the [code of conduct](https://github.com/rubygems/rubygems/blob/master/CODE_OF_CONDUCT.md).

When deploying to an Opsworks instance in AWS (Ubuntu 16.04) our chef cookbook updated rubygems from 2.7.4 to 2.7.5.  After making this change bundler started returning this error...

```
/usr/local/bin/bundle:23:in `load': cannot load such file -- /usr/lib/ruby/gems/2.3.0/gems/bundler-1.16.1/exe/bundle (LoadError)
	from /usr/local/bin/bundle:23:in `<main>'
```
Upgrading rubygems installed bundler in `/var/lib/gems/2.3.0/gems`, but the wrapper script seems to expect to find it in `/usr/lib/ruby/gems/2.3.0/gems/bundler-1.16.1/exe/`

Downgrading to rubygems 2.7.4 via `gem update --system 2.7.4` seems to fix the problem.

Note:  Ruby was installed using the ruby-ng cookbook (https://supermarket.chef.io/cookbooks/ruby-ng), which installs the brightbox-ng ruby package for 2.3.6.  (see https://launchpad.net/~brightbox/+archive/ubuntu/ruby-ng).

**Comment by robuye** [comment]
I'm experiencing the same issue. It appears to be related to the brightbox-ng ruby package. I'm having hard time to find a root cause, but here are steps to reproduce in Docker.

Prepare a working directory:

```
mkdir -p /tmp/rubygems-issue2180
cd /tmp/rubygems-issue2180
```

create the following Dockerfile:

```
FROM ubuntu:16.04

RUN apt-get update && apt-get install -y software-properties-common build-essential

RUN apt-add-repository ppa:brightbox/ruby-ng && \
    apt-get update && \
    apt-get install -y ruby2.4 ruby2.4-dev

RUN gem update --system && \
    gem -v && \
    ruby -v && \
    gem install bundler

CMD ["bundle", "exec", "uptime"]
```

build the image:

```
docker build -t rubygems-issue2180 .
```

Run it to see the error:

```
docker run --rm -it rubygems-issue2180
```

The container is expected to crash with the following error:

```
/usr/local/bin/bundle:23:in `load': cannot load such file -- /usr/lib/ruby/gems/2.4.0/gems/bundler-1.16.1/exe/bundle (LoadError)
        from /usr/local/bin/bundle:23:in `<main>'
```

Run a shell for debugging purposes:

```
docker run --rm -it rubygems-issue2180 bash
```

Downgrading rubygems to 2.7.4 fixes the issue.

Linking the `bundle` from `/var` to `/usr` also fixes the issue:

```
mkdir -p /usr/lib/ruby/gems/2.4.0/gems/bundler-1.16.1/exe/
ln -s /var/lib/gems/2.4.0/gems/bundler-1.16.1/exe/bundle /usr/lib/ruby/gems/2.4.0/gems/bundler-1.16.1/exe/
```

**Comment by hsbt** [comment]
We are hard to fix it deeply related to distribution specific changes like ubuntu and debian. I'm welcome to patch for this issue.

**Comment by robuye** [comment]
Thansk @hsbt, I totally understand!

I think I managed to make some progress. The brightbox package monkeypatches rubygems via `rubygems-integration` package. It includes the following code:

```
unless ENV['DEBIAN_DISABLE_RUBYGEMS_INTEGRATION']

class << Gem
  OPERATING_SYSTEM_DEFAULTS = {
    :ssl_ca_cert => '/etc/ssl/certs/ca-certificates.crt'
  }

  alias :upstream_default_dir :default_dir
  def default_dir
    File.join('/', 'var', 'lib', 'gems', Gem::ConfigMap[:ruby_version])
  end

  alias :upstream_default_bindir :default_bindir
  def default_bindir
    File.join('/', 'usr', 'local', 'bin')
  end

  alias :upstream_default_path :default_path
  def default_path

    # FIXME remove (part of) this after we get rid of ruby 2.1 and 2.2
    extra_path = nil
    if RbConfig::CONFIG['ruby_version'] == '2.1.0'
      extra_path = File.join('/usr/share/rubygems-integration', '2.1')
    elsif RbConfig::CONFIG['ruby_version'] == '2.2.0'
      extra_path = File.join('/usr/share/rubygems-integration', '2.2')
    end

    arch = Gem::ConfigMap[:arch]
    api_version = Gem::ConfigMap[:ruby_version]

    upstream_default_path + [
      "/usr/lib/#{arch}/rubygems-integration/#{api_version}",
      File.join('/usr/share/rubygems-integration', api_version),
      extra_path,
      '/usr/share/rubygems-integration/all'
    ].compact
  end

end

if RUBY_VERSION >= '2.1' then
  class << Gem::BasicSpecification

    alias :upstream_default_specifications_dir :default_specifications_dir
    def default_specifications_dir
      File.join(Gem.upstream_default_dir, "specifications", "default")
    end

  end
end
end
```

I have added the following ENV variable to the dockerfile (anywhere before `gem install bundler`) and it works fine now.

The methods that are being overridden can be found here: 

https://github.com/rubygems/rubygems/blob/master/lib/rubygems/defaults.rb

This monkeypatch appears to follow these sugestions: 

https://github.com/rubygems/rubygems/blob/master/lib/rubygems/commands/environment_command.rb#L64-L66

I still don't understand what change in rubygems 2.7.5 causes this issue :thinking:

**Comment by robuye** [comment]
I have reported the bug to `rubygems-integration` on Launchpad too: https://bugs.launchpad.net/ubuntu/+source/rubygems-integration/+bug/1749226

**Comment by olbrich** [comment]
I've attempted to contact the brightbox ruby-ng developers to see if they have any insight.

**Comment by BRMatt** [comment]
I'm not sure this is related to brightbox's version, I can replicate it with the ruby shipped in xenial's apt repo:

```
FROM ubuntu:xenial

RUN apt-get update && \
  apt-get upgrade -y && \
  apt-get install -y ruby && \
  gem update --system && \
  gem -v && \
  ruby -v && \
  gem install bundler --no-document

CMD /bin/bash
```
Here's the last part of the build output:
```console
RubyGems system software updated
2.7.6
ruby 2.3.1p112 (2016-04-26) [x86_64-linux-gnu]
Successfully installed bundler-1.16.1
1 gem installed
Removing intermediate container d4515383b7e6
 ---> 957804fff3a5
Step 3/3 : CMD /bin/bash
 ---> Running in 9a2e7aed26fa
Removing intermediate container 9a2e7aed26fa
 ---> 1c938399d6d8
Successfully built 1c938399d6d8
```
And here's the error:
```console
$ docker run -it 1c938399d6d8 /bin/bash
root@6856beeffb63:/# bundle
/usr/local/bin/bundle:23:in `load': cannot load such file -- /usr/lib/ruby/gems/2.3.0/gems/bundler-1.16.1/exe/bundle (LoadError)
	from /usr/local/bin/bundle:23:in `<main>'
```

As with the other comments, downgrading to 2.7.4 fixes it:

```console
$ root@9cf2188ac3e8:/# gem update --system  2.7.4
Updating rubygems-update
Fetching: rubygems-update-2.7.4.gem (100%)
Successfully installed rubygems-update-2.7.4
.....
RubyGems system software updated
$ bundle
Don't run Bundler as root. Bundler can ask for sudo if it is needed, and installing your bundle as root will break this application for all non-root users on this machine.
Could not locate Gemfile
```

**Comment by robuye** [comment]
I have received an update from on the package bug report and decided to give it another try to find the root cause.

Here's the reply from Jacob Chae:

> I just looked into this and it looks like the issue is being caused by a change from https://launchpadlibrarian.net/230675272/rubygems-integration_1.9_1.10.diff.gz. The issue is the change to `lib/rubygems/defaults/operating_system.rb` which aliases `default_specifications_dir` to `upstream_default_specifications_dir`. I'm not sure why the change was made but the `upstream_specifications_dir` is what is causing rubygems to look for the gem executables in `/usr/`. If you look at https://github.com/rubygems/rubygems/blob/2ed01b87152cb5ebec3ae934dca50a1609ea2ebd/lib/rubygems/specification.rb#L889, you can see that it uses `default_specifications_dir` to figure out where to load the gems from with `Gem::Specification.load` in `Gem::Specification.each_spec`.

> As I said before I don't know the backstory for this change but changing that, or the default that's being set for `upstream_specifications_dir` seem to be the potential fixes.

I think I have narrowed down the problem. I have revived commits between 2.7.4 and 2.7.5 and with a bit of puts-driven-development I managed to narrow down the issue.

This P/R appears to be introducing the issue: https://github.com/rubygems/rubygems/pull/2169/files#diff-3d2f1e4ce4bf8e4f701d1706efb9f9a1L385

I think this part to be exact:

```diff
-    mkdir_p bundler_spec.bin_dir
-    bundler_spec.executables.each {|e| cp File.join("bundler", bundler_spec.bindir, e), File.join(bundler_spec.bin_dir, e) }
+    bundler_bin_dir = File.join(
+      options[:destdir], Gem.default_dir, 'gems',
+      bundler_spec.full_name, bundler_spec.bindir
+    )
+    mkdir_p bundler_bin_dir
+    bundler_spec.executables.each do |e|
+      cp File.join("bundler", bundler_spec.bindir, e), File.join(bundler_bin_dir, e)
+    end
```


I have restored the removed lines and I could run `bundle` as expected.

I have compared the differences of `bundler_bin_dir` vs `bundler_spec.bin_dir`:

```
bundler_spec.bin_dir: /usr/lib/ruby/gems/2.4.0/gems/bundler-1.16.1/exe
bundler_bin_dir: /var/lib/gems/2.4.0/gems/bundler-1.16.1/exe
```

I'm not sure now what path is correct and what's the global impact of the change. @hsbt could you have another look please? Thanks very much!

**Comment by hsbt** [comment]
@robuye I've merged https://github.com/rubygems/rubygems/pull/2208. I will release 2.7.7 with its fix.

**Comment by robuye** [comment]
Thanks @hsbt, that's awesome!

**Comment by kernow** [comment]
Thanks @hsbt 👍 Do you know when this will be released? I'm waiting on this fix before I can apply the latest security fix update.

**Comment by sj26** [comment]
Here's a workaround we're using for now, in case it's useful to anybody else:

```
# Rubygems now installs bundler, but as a broken "default" gem, so tell it not to
export DONT_USE_BUNDLER_FOR_GEMDEPS=true

# Install latest version of Rubygems
gem update --system

# Install latest version of bundler system-wide
gem install bundler
```

**Comment by Schwad** [comment]
@sj26 we are trying your fix and pray to God it works. Will update.

**Comment by Schwad** [comment]
@sj26 that did not work for us. Right now we are holding out for the 2.7.7 release, @hsbt when can we expect that?

Alternatively, does anyone in the thread know how to update the latest version of ruby gems to point to @hsbt's PR?

**Comment by robuye** [comment]
@Schwad I have been following the README to install master: https://github.com/rubygems/rubygems#installation

Try `ruby setup.rb`

**Comment by Schwad** [comment]
@robuye okay we will try that for installing the latest master. Will update with how we go

**Comment by hsbt** [comment]
As First, You can use bundler and ruby without `gem update --system`. Because I did merge rubygems to the latest stable version of Ruby. So, You can use them with Ruby 2.4.3 or 2.5.0.

> @hsbt when can we expect that?

I'm not a full-time OSS developer. I hard to work to triage a patch and pull-request for releasing 2.7.7. Please wait for its time.

**Comment by Schwad** [comment]
Thank you @hsbt, we are going to try this with Ruby 2.4.3 today. Will update.

**Comment by Schwad** [comment]
We resolved this by upgrading our CI server to Ruby 2.4.3 and pegging bundler to 1.16.0.

Up to a certain point there were references to `/usr/lib/ruby/gems`, and then as of 2.3/2.4/2.5 (somehwere in there) the bundler references changed to `/var/lib/gems`. So we had some references to the var path and some to usr. All running now.

**Comment by MrTh1nk3r** [comment]
sudo gem update --system 2.7.4
gem uninstall bundler
sudo gem install bundler --version '1.16.0'

@Schwad Bundle complete! Thx!

**Comment by Jed-Giblin** [comment]
@MrTh1nk3r Thanks, That worked like a charm. Looks like we will be downgrading for a while

**Comment by axle07** [comment]
With the release of rubygems 2.7.7, this issue can now be resolved with `gem update --system`

**Comment by pboling** [comment]
I am having this bug with latest Rubygems (v2.7.7), latest rvm, and latest Ruby.

```
22:44:38 worker4.1  | /usr/local/rvm/gems/ruby-2.5.1/bin/bundle:23:in `load': cannot load such file -- /usr/local/rvm/rubies/ruby-2.5.1/lib/ruby/gems/2.5.0/gems/bundler-1.16.3/exe/bundle (LoadError)
22:44:38 worker4.1  | 	from /usr/local/rvm/gems/ruby-2.5.1/bin/bundle:23:in `<main>'
22:44:38 worker4.1  | 	from /usr/local/rvm/gems/ruby-2.5.1/bin/ruby_executable_hooks:24:in `eval'
22:44:38 worker4.1  | 	from /usr/local/rvm/gems/ruby-2.5.1/bin/ruby_executable_hooks:24:in `<main>'
```

```
22:47:16 root@smoke-screen:~/cloudsniff (git:fima/upgrade-ruby-2-5+:0025049)  ruby-2.5.1
# gem --version
2.7.7
22:47:20 root@smoke-screen:~/cloudsniff (git:fima/upgrade-ruby-2-5+:0025049)  ruby-2.5.1
# rvm --version
rvm 1.29.4 (master) by Michal Papis, Piotr Kuczynski, Wayne E. Seguin [https://rvm.io]
22:49:04 root@smoke-screen:~/cloudsniff (git:fima/upgrade-ruby-2-5+:0025049)  ruby-2.5.1
# ruby -v
ruby 2.5.1p57 (2018-03-29 revision 63029) [x86_64-linux]
```

**Comment by indirect** [comment]
@pboling have you tried `gem install bundler`, or `gem update --system`? If you have, please provide the output of `gem env` (and `bundle env` if it will run).

**Comment by indirect** [comment]
(also, I just noticed: this ticket is resolved and closed. please open a new ticket, including the relevant information for us to try to reproduce your problem.)
